File: plugins/dynamicHighlights/src/main/java/com/fredplugins/dynamicHighlights/DisplayConfig.java
package com.fredplugins.dynamicHighlights;

import com.fredplugins.dynamicHighlights.model.BufferedImageProvider;
import com.fredplugins.dynamicHighlights.model.FontType;
import com.fredplugins.dynamicHighlights.model.SoundProvider;
import com.fredplugins.dynamicHighlights.model.TextAccent;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import net.runelite.client.ui.FontManager;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

@Getter
@AllArgsConstructor
@EqualsAndHashCode
@ToString
public class DisplayConfig {
    public static final Color DEFAULT_MENU_TEXT_COLOR = Color.decode("#ff9040");

    private final Color textColor;
    private final Color backgroundColor;
    private final Color borderColor;
    private final Boolean hidden;
    private final Boolean showLootbeam;
    private final Boolean showValue;
    private final Boolean compact;
    private final Boolean showDespawn;
    private final Boolean notify;
    private final TextAccent textAccent;
    private final Color textAccentColor;
    private final Color lootbeamColor;
    private final FontType fontType;
    private final Color menuTextColor;
    private final Boolean hideOverlay;

    private final Boolean highlightTile;
    private final Color tileStrokeColor;
    private final Color tileFillColor;

    private final SoundProvider sound;
    private final Integer menuSort;
    private final BufferedImageProvider icon;

    // ideally this would be in EvalDisplayConfig which extends DisplayConfig but that's just more code tbh
    private final List<Integer> evalTrace;

    public DisplayConfig(Color textColor) {
        this.textColor = textColor;
        backgroundColor = null;
        borderColor = null;
        hidden = false;
        showLootbeam = false;
        showValue = false;
        compact = false; // compact is currently a config-only setting and not supported in rs2f
        showDespawn = false;
        notify = false;
        textAccent = null;
        textAccentColor = null;
        lootbeamColor = null;
        fontType = null;
        menuTextColor = null;
        highlightTile = null;
        tileStrokeColor = null;
        tileFillColor = null;
        hideOverlay = null;
        sound = null;
        menuSort = null;
        icon = null;
        evalTrace = new ArrayList<>();
    }

    public static DisplayConfigBuilder builder() {
        return new DisplayConfigBuilder();
    }

    public SoundProvider getSound() {
        return !isHidden() ? sound : null;
    }

    public Color getLootbeamColor() {
        return lootbeamColor != null ? lootbeamColor : textColor;
    }

    public Color getTextColor() {
        return textColor != null ? textColor : Color.WHITE;
    }

    public BufferedImageProvider getIcon() {
        return isCompact() ? new BufferedImageProvider.CurrentItem() : icon;
    }

    public Color getMenuTextColor() {
        if (isHidden()) {
            return DEFAULT_MENU_TEXT_COLOR;
        }
        if (menuTextColor != null) {
            return menuTextColor;
        }
        return textColor != null && !textColor.equals(Color.WHITE) ? textColor : DEFAULT_MENU_TEXT_COLOR;
    }

    public int getMenuSort() {
        return menuSort != null ? menuSort : 0;
    }

    public Font getFont() {
        if (fontType == null || fontType == FontType.NORMAL) {
            return FontManager.getRunescapeSmallFont();
        }
        if (fontType == FontType.LARGER) {
            return FontManager.getRunescapeFont();
        }
        return FontManager.getRunescapeBoldFont();
    }

    public Color getTileStrokeColor() {
        return tileStrokeColor != null ? tileStrokeColor : textColor;
    }

    public boolean isHidden() {
        return hidden != null && hidden;
    }

    public boolean isShowLootbeam() {
        return !isHidden() && showLootbeam != null && showLootbeam;
    }

    public boolean isShowValue() {
        return showValue != null && showValue;
    }

    public boolean isShowDespawn() {
        return showDespawn != null && showDespawn;
    }

    public boolean isNotify() {
        return !isHidden() && notify != null && notify;
    }

    public boolean isHighlightTile() {
        return !isHidden() && highlightTile != null && highlightTile;
    }

    public boolean isHideOverlay() {
        return isHidden() || (hideOverlay != null && hideOverlay);
    }

    public boolean isCompact() {
        return compact != null && compact;
    }

    public DisplayConfig merge(DisplayConfig other) {
        var b = toBuilder();
        if (other.textColor != null) {
            b.textColor(other.textColor);
        }
        if (other.backgroundColor != null) {
            b.backgroundColor(other.backgroundColor);
        }
        if (other.borderColor != null) {
            b.borderColor(other.borderColor);
        }
        if (other.hidden != null) {
            b.hidden(other.hidden);
        }
        if (other.showLootbeam != null) {
            b.showLootbeam(other.showLootbeam);
        }
        if (other.showValue != null) {
            b.showValue(other.showValue);
        }
        if (other.compact != null) {
            b.compact(other.compact);
        }
        if (other.showDespawn != null) {
            b.showDespawn(other.showDespawn);
        }
        if (other.notify != null) {
            b.notify(other.notify);
        }
        if (other.textAccent != null) {
            b.textAccent(other.textAccent);
        }
        if (other.textAccentColor != null) {
            b.textAccentColor(other.textAccentColor);
        }
        if (other.lootbeamColor != null) {
            b.lootbeamColor(other.lootbeamColor);
        }
        if (other.fontType != null) {
            b.fontType(other.fontType);
        }
        if (other.menuTextColor != null) {
            b.menuTextColor(other.menuTextColor);
        }
        if (other.highlightTile != null) {
            b.highlightTile(other.highlightTile);
        }
        if (other.tileStrokeColor != null) {
            b.tileStrokeColor(other.tileStrokeColor);
        }
        if (other.tileFillColor != null) {
            b.tileFillColor(other.tileFillColor);
        }
        if (other.hideOverlay != null) {
            b.hideOverlay(other.hideOverlay);
        }
        if (other.sound != null) {
            b.sound(other.sound);
        }
        if (other.menuSort != null) {
            b.menuSort(other.menuSort);
        }
        if (other.icon != null) {
            b.icon(other.icon);
        }
        return b.build();
    }

    public DisplayConfigBuilder toBuilder() {
        return new DisplayConfigBuilder().textColor(this.textColor).backgroundColor(this.backgroundColor).borderColor(this.borderColor).hidden(this.hidden).showLootbeam(this.showLootbeam).showValue(this.showValue).compact(this.compact).showDespawn(this.showDespawn).notify(this.notify).textAccent(this.textAccent).textAccentColor(this.textAccentColor).lootbeamColor(this.lootbeamColor).fontType(this.fontType).menuTextColor(this.menuTextColor).hideOverlay(this.hideOverlay).highlightTile(this.highlightTile).tileStrokeColor(this.tileStrokeColor).tileFillColor(this.tileFillColor).sound(this.sound).menuSort(this.menuSort).icon(this.icon).evalTrace(this.evalTrace);
    }

    public static class DisplayConfigBuilder {
        private Color textColor;
        private Color backgroundColor;
        private Color borderColor;
        private Boolean hidden;
        private Boolean showLootbeam;
        private Boolean showValue;
        private Boolean compact;
        private Boolean showDespawn;
        private Boolean notify;
        private TextAccent textAccent;
        private Color textAccentColor;
        private Color lootbeamColor;
        private FontType fontType;
        private Color menuTextColor;
        private Boolean hideOverlay;
        private Boolean highlightTile;
        private Color tileStrokeColor;
        private Color tileFillColor;
        private SoundProvider sound;
        private Integer menuSort;
        private BufferedImageProvider icon;
        private List<Integer> evalTrace;

        DisplayConfigBuilder() {
        }

        public DisplayConfigBuilder textColor(Color textColor) {
            this.textColor = textColor;
            return this;
        }

        public DisplayConfigBuilder backgroundColor(Color backgroundColor) {
            this.backgroundColor = backgroundColor;
            return this;
        }

        public DisplayConfigBuilder borderColor(Color borderColor) {
            this.borderColor = borderColor;
            return this;
        }

        public DisplayConfigBuilder hidden(Boolean hidden) {
            this.hidden = hidden;
            return this;
        }

        public DisplayConfigBuilder showLootbeam(Boolean showLootbeam) {
            this.showLootbeam = showLootbeam;
            return this;
        }

        public DisplayConfigBuilder showValue(Boolean showValue) {
            this.showValue = showValue;
            return this;
        }

        public DisplayConfigBuilder compact(Boolean compact) {
            this.compact = compact;
            return this;
        }

        public DisplayConfigBuilder showDespawn(Boolean showDespawn) {
            this.showDespawn = showDespawn;
            return this;
        }

        public DisplayConfigBuilder notify(Boolean notify) {
            this.notify = notify;
            return this;
        }

        public DisplayConfigBuilder textAccent(TextAccent textAccent) {
            this.textAccent = textAccent;
            return this;
        }

        public DisplayConfigBuilder textAccentColor(Color textAccentColor) {
            this.textAccentColor = textAccentColor;
            return this;
        }

        public DisplayConfigBuilder lootbeamColor(Color lootbeamColor) {
            this.lootbeamColor = lootbeamColor;
            return this;
        }

        public DisplayConfigBuilder fontType(FontType fontType) {
            this.fontType = fontType;
            return this;
        }

        public DisplayConfigBuilder menuTextColor(Color menuTextColor) {
            this.menuTextColor = menuTextColor;
            return this;
        }

        public DisplayConfigBuilder hideOverlay(Boolean hideOverlay) {
            this.hideOverlay = hideOverlay;
            return this;
        }

        public DisplayConfigBuilder highlightTile(Boolean highlightTile) {
            this.highlightTile = highlightTile;
            return this;
        }

        public DisplayConfigBuilder tileStrokeColor(Color tileStrokeColor) {
            this.tileStrokeColor = tileStrokeColor;
            return this;
        }

        public DisplayConfigBuilder tileFillColor(Color tileFillColor) {
            this.tileFillColor = tileFillColor;
            return this;
        }

        public DisplayConfigBuilder sound(SoundProvider sound) {
            this.sound = sound;
            return this;
        }

        public DisplayConfigBuilder menuSort(Integer menuSort) {
            this.menuSort = menuSort;
            return this;
        }

        public DisplayConfigBuilder icon(BufferedImageProvider icon) {
            this.icon = icon;
            return this;
        }

        public DisplayConfigBuilder evalTrace(List<Integer> evalTrace) {
            this.evalTrace = evalTrace;
            return this;
        }

        public DisplayConfig build() {
            return new DisplayConfig(this.textColor, this.backgroundColor, this.borderColor, this.hidden, this.showLootbeam, this.showValue, this.compact, this.showDespawn, this.notify, this.textAccent, this.textAccentColor, this.lootbeamColor, this.fontType, this.menuTextColor, this.hideOverlay, this.highlightTile, this.tileStrokeColor, this.tileFillColor, this.sound, this.menuSort, this.icon, this.evalTrace);
        }

        public String toString() {
            return "DisplayConfig.DisplayConfigBuilder(textColor=" + this.textColor + ", backgroundColor=" + this.backgroundColor + ", borderColor=" + this.borderColor + ", hidden=" + this.hidden + ", showLootbeam=" + this.showLootbeam + ", showValue=" + this.showValue + ", compact=" + this.compact + ", showDespawn=" + this.showDespawn + ", notify=" + this.notify + ", textAccent=" + this.textAccent + ", textAccentColor=" + this.textAccentColor + ", lootbeamColor=" + this.lootbeamColor + ", fontType=" + this.fontType + ", menuTextColor=" + this.menuTextColor + ", hideOverlay=" + this.hideOverlay + ", highlightTile=" + this.highlightTile + ", tileStrokeColor=" + this.tileStrokeColor + ", tileFillColor=" + this.tileFillColor + ", sound=" + this.sound + ", menuSort=" + this.menuSort + ", icon=" + this.icon + ", evalTrace=" + this.evalTrace + ")";
        }
    }
}

File: plugins/dynamicHighlights/src/main/scala/com/fredplugins/dynamicHighlights/DisplayConfig.scala
package com.fredplugins.dynamicHighlights
import com.fredplugins.dynamicHighlights.model.BufferedImageProvider
import com.fredplugins.dynamicHighlights.model.FontType
import com.fredplugins.dynamicHighlights.model.SoundProvider
import com.fredplugins.dynamicHighlights.model.TextAccent
import net.runelite.client.ui.FontManager
import net.runelite.client.util.ImageUtil

import java.awt.Color
import java.awt.Font
import java.awt.image.BufferedImage
import scala.jdk.CollectionConverters.*
import scala.jdk.OptionConverters.*
import scala.jdk.StreamConverters.*
import scala.util.chaining.*
import scala.util.{Random, Try}
import scala.compiletime.uninitialized
import scala.swing.Color


case class DisplayConfig(
	showValue:Option[Boolean],
	showDespawn: Option[Boolean],
	textColor: Option[Color],
	backgroundColor: Option[Color],
	borderColor: Option[Color],
	menuTextColor: Option[Color],
	tileStrokeColor: Option[Color],
	tileFillColor: Option[Color],
	highlightTile: Option[Boolean],
	lootbeamColor: Option[Color],
	showLootbeam: Option[Boolean],
	accent: Option[TextAccent],
	accentColor: Option[Color],
	fontStyle: Option[FontType],
	icon: Option[BufferedImageProvider],
	sound: Option[SoundProvider],
	menuSort: Option[Int],
	hideOverlay: Option[Boolean],
	hideAll: Option[Boolean],
	notifyPlayer: Option[Boolean],
	trace: Seq[Int]
) {
	def getTrace(): Array[Int] = trace.toArray
	def getSound: SoundProvider = if(!isHidden) sound.orNull else null
	def getLootbeamColor: Color = lootbeamColor.getOrElse(getTextColor)
	def getTextColor: Color = textColor.getOrElse(Color.WHITE)
	def getIcon: BufferedImageProvider = icon.orNull

	def getMenuTextColor: Color = {
		Option.when(!isHidden)(
			menuTextColor.orElse(textColor)
		).flatten.getOrElse(Color.decode("#ff9040"))
	}

	def getMenuSort: Int = menuSort.getOrElse(0)

	def getFont: Font = {
		fontStyle.collect {
			case FontType.LARGER => FontManager.getRunescapeFont
			case FontType.BOLD => FontManager.getRunescapeBoldFont
		}.getOrElse(FontManager.getRunescapeSmallFont)
	}

	def getTileStrokeColor: Color =tileStrokeColor.getOrElse(getTextColor)

	def isHidden: Boolean = hideAll.isDefined && hideAll.get
	def isShowLootbeam: Boolean = Option.when(!isHidden)(showLootbeam).flatten.exists(_ == true)
	def isShowValue: Boolean = showValue.isDefined && showValue.get
	def isShowDespawn: Boolean = showDespawn.exists(_== true)
	def isNotify: Boolean = Option.when(!isHidden)(notifyPlayer).flatten.exists(_ == true)

	def isHighlightTile: Boolean = Option.when(!isHidden)(highlightTile).flatten.exists(_ == true)
	def isHideOverlay: Boolean = {
		if(hideOverlay.isDefined && hideOverlay.get) {
			return true
		}
		if(hideAll.isDefined && hideAll.get) {
			return true
		}
		return false
//		val b1 = isHidden || hideOverlay.contains(true)
//		val b2 = Option.when(!isHidden)(hideOverlay).flatten.exists(_ == true)
//		if(b1 != b2) {
//			println(s"isHidden=${isHidden}, hideOverlay=${hideOverlay}, b1=${b1}, b2=${b2}")
//		}
//		b1
	}
}

object DisplayConfig {

	class DisplayConfigBuilder(initial: DisplayConfig) {
		private var data: DisplayConfig = initial
		def build: DisplayConfig = data.copy()

		def withTextColor(c: Color): this.type = {
			data = data.copy(textColor = Option(c))
			this
		}
		def withBackgroundColor(c: Color): this.type = {
			data = data.copy(backgroundColor = Option(c))
			this
		}
		def withBorderColor(c: Color): this.type = {
			data = data.copy(borderColor = Option(c))
			this
		}
		def withHidden(c: Boolean): this.type = {
			data = data.copy(hideAll = Option(c))
			this
		}

		def withShowLootbeam(c: Boolean): this.type = {
			data = data.copy(showLootbeam = Option(c))
			this
		}

		def withShowValue(c: Boolean): this.type = {
			data = data.copy(showValue = Option(c))
			this
		}
		def withShowDespawn(c: Boolean): this.type = {
			data = data.copy(showDespawn = Option(c))
			this
		}
		def withNotify(c: Boolean): this.type = {
			data = data.copy(notifyPlayer = Option(c))
			this
		}

		def withTextAccent(c: TextAccent): this.type = {
			data = data.copy(accent = Option(c))
			this
		}
		def withAccentColor(c: Color): this.type = {
			data = data.copy(accentColor = Option(c))
			this
		}

		def withLootbeamColor(c: Color): this.type = {
			data = data.copy(lootbeamColor = Option(c))
			this
		}
		def withFontType(c: FontType): this.type = {
			data = data.copy(fontStyle = Option(c))
			this
		}
		def withMenuTextColor(c: Color): this.type = {
			data = data.copy(menuTextColor = Option(c))
			this
		}
		def withHighlightTile(c: Boolean): this.type = {
			data = data.copy(highlightTile = Option(c))
			this
		}
		def withTileStrokeColor(c: Color): this.type = {
			data = data.copy(tileStrokeColor = Option(c))
			this
		}

		def withTileFillColor(c: Color): this.type = {
			data = data.copy(tileFillColor = Option(c))
			this
		}
		def withHideOverlay(c: Boolean): this.type = {
			data = data.copy(hideOverlay = Option(c))
			this
		}
		def withSound(c: SoundProvider): this.type = {
			data = data.copy(sound = Option(c))
			this
		}
		def withIcon(c: BufferedImageProvider): this.type = {
			data = data.copy(icon = Option(c))
			this
		}
		def withMenuSort(i: Int): this.type = {
			data = data.copy(menuSort = Option(i))
			this
		}
		def withTrace(i: Int): this.type  = {
			data = data.copy(trace = data.trace.appended(i))
			this
		}
	}

	val empty: DisplayConfig = DisplayConfig(
		Option.empty, Option.empty, Option.empty, Option.empty,
		Option.empty, Option.empty, Option.empty, Option.empty,
		Option.empty, Option.empty, Option.empty, Option.empty,
		Option.empty, Option.empty, Option.empty, Option.empty,
		Option.empty, Option.empty, Option.empty, Option.empty,
		Seq.empty
	)
	def emptyBuilder: DisplayConfigBuilder = empty.builder

	extension (a:DisplayConfig) {
		def builder: DisplayConfigBuilder = DisplayConfigBuilder(a)
		def merge(b: DisplayConfig): DisplayConfig = {
			var mut = a.builder
			b.showValue.foreach(mut.withShowValue(_))
			b.showDespawn.foreach(mut.withShowDespawn(_))
			b.textColor.foreach(mut.withTextColor(_))
			b.backgroundColor.foreach(mut.withBackgroundColor(_))
			b.borderColor.foreach(mut.withBorderColor(_))
			b.menuTextColor.foreach(mut.withMenuTextColor(_))
			b.tileStrokeColor.foreach(mut.withTileStrokeColor(_))
			b.tileFillColor.foreach(mut.withTileFillColor(_))
			b.highlightTile.foreach(mut.withHighlightTile(_))
			b.lootbeamColor.foreach(mut.withLootbeamColor(_))
			b.showLootbeam.foreach(mut.withShowLootbeam(_))
			b.accent.foreach(mut.withTextAccent(_))
			b.accentColor.foreach(mut.withAccentColor(_))
			b.fontStyle.foreach(mut.withFontType(_))
			b.icon.foreach(mut.withIcon(_))
			b.sound.foreach(mut.withSound(_))
			b.menuSort.foreach(mut.withMenuSort(_))
			b.hideOverlay.foreach(mut.withHideOverlay(_))
			b.hideAll.foreach(mut.withHidden(_))
			b.notifyPlayer.foreach(mut.withNotify(_))
			mut.build
		}
	}

	def apply(txtColor: Color): DisplayConfig = {
		empty.builder.withTextColor(txtColor).withHidden(false).withShowLootbeam(false).withShowValue(false).withShowDespawn(false).withNotify(false).build
	}
}
